pkg/cmd/pulumi/crypto.go

```diff
@@ -56,31 +57,50 @@ func getStackSecretsManager(s backend.Stack) (secrets.Manager, error) {
 		return nil, err
 	}
 
-	sm, err := func() (secrets.Manager, error) {
-		configFile, err := getProjectStackPath(s)
-		if err != nil {
-			return nil, err
-		}
-
-		//nolint:goconst
-		if ps.SecretsProvider != passphrase.Type && ps.SecretsProvider != "default" && ps.SecretsProvider != "" {
-			return cloud.NewCloudSecretsManager(
-				s.Ref().Name(), configFile, ps.SecretsProvider, false /* rotateSecretsProvider */)
-		}
-
-		if ps.EncryptionSalt != "" {
-			return passphrase.NewPromptingPassphraseSecretsManager(
-				s.Ref().Name(), configFile, false /* rotateSecretsProvider */)
-		}
-
-		return s.DefaultSecretManager(configFile)
-	}()
+	oldConfig := deepcopy.Copy(ps).(*workspace.ProjectStack)
+
+	var sm secrets.Manager
+	//nolint:goconst
+	if ps.SecretsProvider != passphrase.Type && ps.SecretsProvider != "default" && ps.SecretsProvider != "" {
+		sm, err = cloud.NewCloudSecretsManager(
+			ps, ps.SecretsProvider, false /* rotateSecretsProvider */)
+	} else if ps.EncryptionSalt != "" {
+		sm, err = passphrase.NewPromptingPassphraseSecretsManager(
+			ps, false /* rotateSecretsProvider */)
+	} else {
+		sm, err = s.DefaultSecretManager(ps)
+	}
 	if err != nil {
 		return nil, err
 	}
+
+	// Handle if the configuration changed any of EncryptedKey, etc
+	if err := saveProjectStackAfterSecretManger(s, oldConfig, ps); err != nil {
+		return nil, err
+	}
 	return stack.NewCachingSecretsManager(sm), nil
 }
 
+func saveProjectStackAfterSecretManger(stack backend.Stack,
```

Comment:
Deep copy and compare on just those fields feels like it'll break easily in the future.

This is probably fine right now, but would a longer term fix here be if the individual secrets manager constructors did one of the following?

- return whether they modified ProjectStack (e.g. `func NewCloudSecretsMangaer(...) (sm secrets.Manager,  changed bool, err error)`
- return ProjectStackSecretSettings (or something) that the caller merges into ProjectStack
- mark the ProjectStack dirty with a `func (*ProjectStack) SetDirty(bool)` where they normally would have saved, and the caller inspects it (`func (*ProjectStack) Dirty() bool`) to decide if it needs to save

Reply:
> return ProjectStackSecretSettings (or something) that the caller merges into ProjectStack

I'm planning on doing something like that. This current system clearly won't work if we have user defined plugin secret systems.